File: 2197-890-1123-lowest-common-ancestor-of-deepest-leaves/2197-890-1123-lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def mxdepth(root):
            if not root:
                return 0
            return max(mxdepth(root.left),mxdepth(root.right))+1
        def dfs(root,d,mxd):
            if not root:
                return None
            if d == mxd:
                return root
            l = dfs(root.left,d+1,mxd)
            r = dfs(root.right,d+1,mxd)
            if l and r:
                return root
            return l if l else r
        return dfs(root,1,mxdepth(root))

File: 2197-890-1123-lowest-common-ancestor-of-deepest-leaves/2197-890-1123-lowest-common-ancestor-of-deepest-leaves.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var lcaDeepestLeaves = function(root) {
    function mxdepth(root) {
        if (!root) {
            return 0;
        }
        return Math.max(mxdepth(root.left), mxdepth(root.right)) + 1;
    }
    
    function dfs(root, d, mxd) {
        if (!root) {
            return null;
        }
        if (d === mxd) {
            return root;
        }
        
        const l = dfs(root.left, d + 1, mxd);
        const r = dfs(root.right, d + 1, mxd);
        
        if (l && r) {
            return root;
        }
        return l ? l : r;
    }
    
    return dfs(root, 1, mxdepth(root));
};